src/invite-api.js

```diff
@@ -1,236 +1,440 @@
 // @ts-check
 import { TypedEmitter } from 'tiny-typed-emitter'
 import { InviteResponse_Decision } from './generated/rpc.js'
-import { projectKeyToId, projectKeyToPublicId } from './utils.js'
+import { assert, keyToId } from './utils.js'
+import timingSafeEqual from './lib/timing-safe-equal.js'
+import { promiseWithResolvers } from './ponyfills.js'
 
 /**
- * @typedef {Object} InviteApiEvents
+ * Internally, we typically use the `Invite` type from the protobuf. We also use
+ * an external type for public consumers.
  *
- * @property {(info: { projectId: string, projectName?: string, peerId: string, roleName: string, roleDescription: string | undefined, invitorName: string }) => void} invite-received
+ * @internal
+ * @typedef {import('./generated/rpc.js').Invite} InviteInternal
+ */
+
+/**
+ * @internal
+ * @typedef {import('./generated/rpc.js').ProjectJoinDetails} ProjectJoinDetails
  */
 
 /**
- * @template K
- * @template V
- * @extends {Map<K, Set<V>>}
+ * @typedef {object} Invite
+ * @prop {string} inviteId
+ * @prop {string} projectName
+ * @prop {string} [roleName]
+ * @prop {string} [roleDescription]
+ * @prop {string} [invitorName]
  */
-class MapOfSets extends Map {
-  /** @param {K} key */
-  get(key) {
-    const existing = super.get(key)
-    if (existing) return existing
-    /** @type {Set<V>} */
-    const set = new Set()
-    super.set(key, set)
-    return set
+
+/**
+ * Manage pending invite state.
+ */
+class PendingInvites {
+  /**
+   * @internal
+   * @typedef {object} PendingInvite
+   * @prop {string} peerId
+   * @prop {InviteInternal} invite
+   * @prop {boolean} isAccepting
+   */
+
+  /** @type {Map<string, PendingInvite>} */
+  #byInviteId = new Map()
+
+  /** @type {Map<string, PendingInvite[]>} */
+  #byProjectPublicId = new Map()
+
+  /**
+   * @returns {Iterable<PendingInvite>} the pending invites, in insertion order
+   */
+  invites() {
+    return this.#byInviteId.values()
+  }
+
+  /**
+   * @param {PendingInvite} pendingInvite
+   * @throws if adding a duplicate invite ID
+   * @returns {void}
+   */
+  add(pendingInvite) {
+    const { invite } = pendingInvite
+    const { inviteId, projectPublicId } = invite
+
+    const inviteIdString = keyToId(inviteId)
+    assert(!this.#byInviteId.has(inviteIdString), 'Added duplicate invite')
+    this.#byInviteId.set(inviteIdString, pendingInvite)
+
+    const existingForProject = this.#byProjectPublicId.get(projectPublicId)
+    if (existingForProject) {
+      existingForProject.push(pendingInvite)
+    } else {
+      this.#byProjectPublicId.set(projectPublicId, [pendingInvite])
+    }
+  }
+
+  /**
+   * @param {Buffer} inviteId
+   * @returns {void}
+   */
+  markAccepting(inviteId) {
```

Comment:
I'm often tempted to just mutate objects for something like this, but I should probably ignore those evil thoughts

Reply:
I considered that too...the siren call of mutation beckons us both. I opted to avoid mutation here because these objects are used outside this class, but I don't feel strongly.

Reply:
That seems like a good reason to me.